File: packages/capture-structuring/src/contracts.ts
import type {
  CaptureBlockV1,
  CaptureEngineV1,
  RawCaptureSegmentV1,
  RawCaptureV1,
} from '@gx-capture/capture-contracts';

/** JSON Schema object supplied to an LLM generation provider. */
export type StructuringSchema = Readonly<Record<string, unknown>>;

/** JSON response accepted from a host-owned LLM callback. */
export type StructuringCandidate = Uint8Array | string;

/** Prompt payload sent to the host-owned LLM generation callback. */
export type StructuringBatchPrompt = Readonly<{
  /** Human-readable instruction for the bounded semantic task. */
  instruction: string;
  /** Optional target language, or `null` for identity structuring. */
  targetLanguage: string | null;
  /** Bounded raw segment payloads made available to the model. */
  rawSegments: readonly Readonly<Record<string, unknown>>[];
}>;

/** The minimal semantic fields that may cross the untrusted LLM boundary. */
export type CaptureSemanticBlockV1 = Readonly<{
  /** Stable identifier binding the semantic result to one raw segment. */
  sourceSegmentId: string;
  /** Semantic block classification selected by the model. */
  type: CaptureBlockV1['type'];
  /** Translated text; omitted for identity structuring. */
  targetText?: string;
}>;

/** Host callback used by the brain-agnostic SDK to obtain one JSON candidate. */
export type LlmGenerate = (
  prompt: StructuringBatchPrompt,
  schema: StructuringSchema,
) => StructuringCandidate | PromiseLike<StructuringCandidate>;

/** Error raised when an LLM candidate cannot be safely projected. */
export class StructuringValidationError extends Error {
  /** Machine-readable validation issue details, when available. */
  readonly issues: readonly Readonly<Record<string, unknown>>[];

  /**
   * Creates a structuring validation error.
   *
   * @param message Human-readable failure description.
   * @param issues Optional locations and details for the failure.
   */
  constructor(
    message: string,
    issues: readonly Readonly<Record<string, unknown>>[] = [],
  ) {
    super(message);
    this.name = 'StructuringValidationError';
    this.issues = issues;
  }
}

/** Token-budget accounting for one host LLM request. */
export interface StructuringBatchPlan {
  /** Raw segments included in this request, in source order. */
  readonly segments: readonly RawCaptureSegmentV1[];
  /** Estimated input token count including fixed prompt/schema overhead. */
  readonly inputTokens: number;
  /** Estimated output token count including fixed response overhead. */
  readonly outputTokens: number;
}

/** Inputs required to structure one raw capture through a host-owned LLM. */
export interface StructureCaptureOptions {
  /** Canonical raw capture whose provenance must be preserved. */
  readonly raw: RawCaptureV1;
  /** Host callback that performs the actual LLM generation. */
  readonly llmGenerate: LlmGenerate;
  /** Trusted identity of the engine that produced the structured result. */
  readonly structuringEngine: CaptureEngineV1;
  /**
   * Optional semantic response schema. Defaults to the canonical Capture
   * schema; Ollama hosts may pass an Ollama-compatible schema explicitly.
   */
  readonly schema?: StructuringSchema;
  /** Optional translation target; omission selects identity structuring. */
  readonly targetLanguage?: string;
  /** Trusted completion timestamp assigned by the host. */
  readonly completedAt: string;
  /** Optional maximum context budget for each request. */
  readonly numCtx?: number;
  /** Optional maximum generation budget for each request. */
  readonly numPredict?: number;
}

File: packages/capture-structuring-python/src/capture_structuring/contracts.py
"""Public types and validated semantic models for the Python SDK."""

from __future__ import annotations

from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints
from pydantic.alias_generators import to_camel

NonEmptyString = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
"""String value that must contain at least one non-whitespace character."""

CaptureText = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=2_000_000),
]
"""Capture text constrained to the canonical raw/target text size limit."""

type WireObject = dict[str, object]
"""JSON-compatible object used at the SDK wire boundary."""

type WireInput = Mapping[str, object] | object
"""Mapping or Pydantic-like object accepted by the SDK wire boundary."""

type StructuringSchema = Mapping[str, object]
"""JSON Schema object supplied to an LLM generation provider."""

type StructuringCandidate = bytes | bytearray | str
"""JSON candidate representation accepted from a host-owned LLM callback."""

type LlmGenerate = Callable[
    [Mapping[str, object], StructuringSchema],
    StructuringCandidate | Awaitable[StructuringCandidate],
]
"""Host callback that turns one prompt/schema pair into JSON bytes."""


class StructuringValidationError(ValueError):
    """Raised when a host response cannot be safely projected into a document."""

    def __init__(self, message: str, *, issues: list[dict[str, object]] | None = None) -> None:
        """Create an error with optional machine-readable validation issues."""

        super().__init__(message)
        self.issues = issues or []


class _SemanticModel(BaseModel):
    """Base model enforcing camelCase JSON aliases and forbidden extras."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        extra="forbid",
        populate_by_name=True,
        str_strip_whitespace=True,
    )


class CaptureSemanticBlock(_SemanticModel):
    """The only block fields a host LLM is allowed to return."""

    source_segment_id: NonEmptyString
    type: Literal["heading", "paragraph", "list-item", "table", "quote", "transcript"]
    target_text: CaptureText | None = None


class CaptureIdentitySemanticBlock(_SemanticModel):
    """Identity-mode semantics, which must not include a target-text echo."""

    source_segment_id: NonEmptyString
    type: Literal["heading", "paragraph", "list-item", "table", "quote", "transcript"]


class CaptureBlockBatch(_SemanticModel):
    """Translated semantic blocks returned for one bounded host request."""

    blocks: list[CaptureSemanticBlock] = Field(min_length=1)


class CaptureIdentityBlockBatch(_SemanticModel):
    """Identity semantic blocks returned for one bounded host request."""

    blocks: list[CaptureIdentitySemanticBlock] = Field(min_length=1)


class _ValidatedSemanticBlock:
    """Structural view shared by translated and identity semantic models."""

    source_segment_id: str
    type: Literal["heading", "paragraph", "list-item", "table", "quote", "transcript"]


@dataclass(frozen=True, slots=True)
class StructuringBatchPlan:
    """Token-budget accounting for one host LLM request."""

    segments: tuple[object, ...]
    input_tokens: int
    output_tokens: int


__all__ = [
    "CaptureBlockBatch",
    "CaptureIdentityBlockBatch",
    "CaptureIdentitySemanticBlock",
    "CaptureSemanticBlock",
    "CaptureText",
    "LlmGenerate",
    "NonEmptyString",
    "StructuringBatchPlan",
    "StructuringCandidate",
    "StructuringSchema",
    "StructuringValidationError",
    "WireInput",
    "WireObject",
]
